core/state_transition.go

```diff
@@ -248,15 +247,21 @@ func (st *StateTransition) TransitionDb() (ret []byte, requiredGas, usedGas *big
 		// sufficient balance to make the transfer happen. The first
 		// balance transfer may never fail.
 		if vmerr == vm.ErrInsufficientBalance {
-			return nil, nil, nil, vmerr
+			return nil, nil, nil, false, vmerr
 		}
 	}
 	requiredGas = new(big.Int).Set(st.gasUsed())
 
 	st.refundGas()
 	st.state.AddBalance(st.evm.Coinbase, new(big.Int).Mul(st.gasUsed(), st.gasPrice))
 
-	return ret, requiredGas, st.gasUsed(), err
+	// Derive transaction execution status from the returned error.
+	// Notes, there is no possible to return an ErrDepth error from Call Or Create.
+	if vmerr != nil && (evm.ChainConfig().IsHomestead(evm.BlockNumber) || err != vm.ErrCodeStoreOutOfGas) {
+		failed = true
+	}
+
+	return ret, requiredGas, st.gasUsed(), failed, err
```

Comment:
This whole block can be replaces imho with:

`return ret, requiredGas, st.gasUsed(), vmerr != nil, nil`

i.e. the `failed` flag is only ever used upward from Metropolis, so we don't really care if it's true or false in Frontier.